platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,278 @@
+
+#### Pain points
+
+##### Content Writers
+
+Content writers want to quickly draft, publish, and deploy content. However, `vets-website` is only deployed to production once per day. To allow content writers to move faster, there is a [partial deploy](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/workflow/deploy/#partial-deploy--static-page-changes-only), which only includes static page changes (`vagov-content` and Drupal).
+
+No accessibility checking is performed on the content-only deploys. That means that content writers don't have a proactive way to learn if their content includes accessibility errors. Adding accessibility checking to the CMS is on the roadmap, but it doesn't currently exist.
+
+##### VA.gov Users
+
+To enable a faster publishing workflow, all tests (except link checking) are intentionally skipped during content-only deploys. That creates an opportunity for end users to encounter accessibility errors. Additionally, inconsistencies between how the CMS handles redirects and how `vets-website` handles redirects, it's possible for users to encounter broken links in production.
+
+##### FE Tools team
+
+Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with the CMS team.
+
+##### Anybody that is trying to ship code for the VA.gov front end
+
+If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.
+
+## High Level Design
+
+As part of the [build separation wor](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2719)k, two separate pipelines will be created for for built code to be deployed to S3 to prevent failures in one from blocking between CMS and application deployments.
+
+The Metalsmith static content build will be extracted to the `content-build` repo and deleted from `vets-website`, and its deployment pipeline. Webpack will handle the application build for the `vets-website` repo, and Metalsmith will handle the static content build for the `content-build` repo.
+
+The validation of static content will take place in a scheduled job that runs every workday instead of ???. Since the broken link check happens during the Metalsmith build, and the accessibility check relies on the output of the content build, the scheduled job will run a script to produce a build that only has the steps necessary for content validation. That build will only be used for reporting broken links and accessibility errors, and will have no impact on deploys. Those reports will be delivered to the #cms-team channel via existing Slack integrations.
+
+## Specifics
+
+### Detailed Design
```

Comment:
this makes sense as a starting point, but I think needs a _lot_ more detail.

each of the sentences you have below should probably have their own heading describing that decision and explanations about how to implement each, and maybe describing how they affect the pain points. but at the moment there isn't enough described to feel like this really communicates what's happening.

some initial questions that come to mind as examples of what detail this still needs:

> Using the current content build script as a reference, we will write another build script that only includes the steps necessary for the content validation. 

* Does that mean you'll remove content validation from the current content build script?
* What runs this new build script?
* What happens when this new build script detects invalid content?
* Will invalid content detection block a build or deployment?

> We will reuse the existing broken link check from the current content build. 

* does that mean just the checking logic? what about how reporting is done?

> And using the existing accessibility tests as a reference, we will configure the accessibility tests to not stop on failure so we can identify all the accessibility errors.

* Wait, are you using the accessibility tests as a reference for other work? Or changing the accessibility tests? This sounds like both, but to me those are mutually exclusive things
* What does "not stop on failure" mean --- just making sure we check for _all_ errors instead of stopping at the first error? Or does it mean not stopping a build/deploy?


To be clear, these are all separate questions that I feel like the design doc doesn't answer currently.

Reply:
A few other questions I have reading this:

* Where will this script run? (Jenkins? CircleCI? somewhere else?)
* How often will it run? (you said "every workday" above, does that mean once/day? if it's just giving us notifications, should we run it more frequently?)
* What content will it run against? Fetched directly from Drupal? Running against what's live on the website?

Reply:
Thanks for the extra detail! I feel much more able to understand your proposal now.

I didn't find an answer to this question of mine though:

> What content will it run against? Fetched directly from Drupal? Running against what's live on the website?

Reply:
> Thanks for the extra detail! I feel much more able to understand your proposal now.
> 
> I didn't find an answer to this question of mine though:
> 
> > What content will it run against? Fetched directly from Drupal? Running against what's live on the website?

@gunsch, sorry for missing that question. To answer your question, we will first try to pull the latest content from Drupal. If we are unable to pull the latest content from Drupal, we will use the cached Drupal content.

I've updated the design doc to include that answer in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/6077200b080141dd0b7946cd6060fb54ff9e42ee.